src/librbd/operation/MigrateRequest.cc

```diff
@@ -68,25 +101,129 @@ bool MigrateRequest<I>::should_complete(int r) {
   if (r < 0) {
     lderr(cct) << "encountered error: " << cpp_strerror(r) << dendl;
   }
+
   return true;
 }
 
 template <typename I>
-void MigrateRequest<I>::send_op() {
+void MigrateRequest<I>::copyup_objects() {
   I &image_ctx = this->m_image_ctx;
-  assert(image_ctx.owner_lock.is_locked());
   CephContext *cct = image_ctx.cct;
-  ldout(cct, 5) << dendl;
+  assert(image_ctx.owner_lock.is_locked());
+
+  if (image_ctx.size == 0) {
+    deep_copy_objects();
+    return;
+  }
 
+  ldout(cct, 5) << "from 0 to " << m_image_size_objects << dendl;
+
+  auto ctx = create_context_callback<
+    MigrateRequest<I>, &MigrateRequest<I>::handle_copyup_objects>(this);
+
+  assert(m_total_objects >= m_image_size_objects);
+  m_copyup_prog_ctx = new CopyupProgressContext(m_prog_ctx, m_total_objects);
   typename AsyncObjectThrottle<I>::ContextFactory context_factory(
     boost::lambda::bind(boost::lambda::new_ptr<C_MigrateObject<I> >(),
-      boost::lambda::_1, &image_ctx, m_snapc, boost::lambda::_2));
+      boost::lambda::_1, &image_ctx, image_ctx.snapc, boost::lambda::_2));
   AsyncObjectThrottle<I> *throttle = new AsyncObjectThrottle<I>(
-    this, image_ctx, context_factory, this->create_callback_context(), &m_prog_ctx,
-    0, m_overlap_objects);
+    this, image_ctx, context_factory, ctx, m_copyup_prog_ctx, 0,
+    m_image_size_objects);
   throttle->start_ops(image_ctx.concurrent_management_ops);
 }
 
+template <typename I>
+void MigrateRequest<I>::handle_copyup_objects(int r) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 5) << "r=" << r << dendl;
+
+  delete m_copyup_prog_ctx;
+  m_copyup_prog_ctx = nullptr;
+
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to copyup objects: " << cpp_strerror(r) << dendl;
+    this->complete(r);
+    return;
+  }
+
+  deep_copy_objects();
+}
+
+template <typename I>
+void MigrateRequest<I>::deep_copy_objects() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+
+  if (m_image_size_objects == m_total_objects) {
+    this->complete(0);
+    return;
+  }
+
+  assert(m_total_objects > m_image_size_objects);
+
+  ldout(cct, 5) << "from " << m_image_size_objects << " to " << m_total_objects
+                << dendl;
+
+  assert(image_ctx.parent != nullptr);
```

Comment:
Nit: not holding ```ImageCtx::parent_lock```

Reply:
This is an interesting question. I even wanted to leave a comment here. Ok, I can hold a lock when checking `image_ctx.parent != nullptr`, but how can I make sure `image_ctx.parent` has not gone during `deep_copy::ImageCopyRequest`?

Reply:
You could register a dummy ```io::AsyncOperation``` within ```image_ctx.parent->async_ops``` -- those are flushed prior to closing the parent. It would result in the refresh "hanging" until the migration completes, but I really wouldn't expect the migration source to be yanked while a migration is in-progress.

Reply:
... or just re-implement the per-object throttled loop here to construct / send ```deep_copy::ObjectCopyRequest```s so that you can test the existence of the parent for each individual object and abort.